Prompt user to remove application credentials when deleting the integration configuration

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->

Prompt user to remove application credentials when deleting the integration configuration. This is to avoid user confusion where they delete the integration thinking they have a clean slate, but still have application credentials hidden (they don't know where to look in the overflow menu yet).

<img width="542" alt="Screen Shot 2022-09-04 at 2 03 38 PM" src="https://user-images.githubusercontent.com/6026418/188333475-a7cf8699-6fca-4445-84b1-0616d413cc91.png">


Related Core PR: https://github.com/home-assistant/core/pull/74825


## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: 
- This PR is related to issue or discussion: https://github.com/home-assistant/core/issues/74552 https://github.com/home-assistant/home-assistant.io/issues/23330
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

allenporter committed:
Prompt user to delete application credentials at integration delete

allenporter committed:
Prompt user to delete application credentials at integration delete

allenporter committed:
Cleanup text and alerts for application credentails removal

robloh commented:
As an end user I would be puzzled by this dialog and what it's asking of me. What are the ramifications of picking either cancel or remove and how does it affect "my" future?

I think it needs to explain what the differences are between the 2 choices or just always remove credentials if they're no longer needed?

allenporter committed:
Update the variable name from core pr.

allenporter commented:
> As an end user I would be puzzled by this dialog and what it's asking of me. What are the ramifications of picking either cancel or remove and how does it affect "my" future?
> 
> I think it needs to explain what the differences are between the 2 choices or just always remove credentials if they're no longer needed?

Thanks that is very helpful feedback.

So, let me see if i can explain the choice here, then we can discuss the wording.

- If you remove them, then you are prompted to re-add them next time the integration is setup.  Maybe you have them handy or if not then you'll see a link of where to go to find them on the google cloud side
- If you leave them, they will be automatically re-used when setting up the integration so you don't have to re-enter them again which is handy.

There is a desire to have integrations and the credentials operate separately, certainly at the schema level, but it does create this user tension of having to know they MAY operate independently even if they don't care, so as an intermediate step I at least want to raise awareness these exist.  Note that you can also use the credentials for multiple config entries.


So, maybe we want to rename "Cancel" to "Keep".

Maybe something like this is better?
> Would you like to keep or remove them? If you keep them, they will automatically be used when setting up the integration again in the future. If you remove them, you will be prompted to re-enter the credentials when setting up the integration again in the future.

I'd also love input on improvements.

spacegaier commented:
Two initial ideas: 

1. Add a link (labeled "Help me decide"?) at the bottom of the dialog pointing to a FAQ in the HA docs where we explain what you wrote in your comment above.
2. Or if we do not want the online dependency to HA, perhaps an expandable/collapsible section in the dialog where then the explanation is shown. That way the dialog stays nice/small/simple, but offers more context if needed.

spacegaier commented:
I also was recently looking at this ticket https://github.com/home-assistant/frontend/issues/9773 that I wanted to tackle at some point, but this credentials dialog here made me pause: "Delete" for an integration is fine, but in the case of the credentials we are technically only "removing" them from HA as the credentials themselves (configured on some other platform outside HA) of course stay valid and are not "deleted" or invalidated by that action. 

So not sure if that is a distinction we have to care about?

allenporter commented:
> I also was recently looking at this ticket #9773 that I wanted to tackle at some point, but this credentials dialog here made me pause: "Delete" for an integration is fine, but in the case of the credentials we are technically only "removing" them from HA as the credentials themselves (configured on some other platform outside HA) of course stay valid and are not "deleted" or invalidated by that action.
> 
> So not sure if that is a distinction we have to care about?

Hi, my interpretation is that you are saying the text should say "remove" when referring to application credentials which is what it is doing now. Am i understanding that right? or are you pointing out that this is saying remove which is inconsistent with delete?

spacegaier commented:
I am saying that your currently chosen wording makes sense to me, but I remember that there is the linked issue where a rename was proposed, but I just realized, that this ticket is purely talking about "entities", so probably does not really apply here. I have pinged @matthiasdebaat to chime in to confirm.

allenporter committed:
Use new websocket api for checking if credentials are in use

allenporter committed:
Revert delete config entry change

allenporter commented:
> I am saying that your currently chosen wording makes sense to me, but I remember that there is the linked issue where a rename was proposed, but I just realized, that this ticket is purely talking about "entities", so probably does not really apply here. I have pinged @matthiasdebaat to chime in to confirm.

OK thanks for pointing this out. Definitely happy to defer to the experts here and follow whatever pattern is recommended.

allenporter committed:
Update application credentials wording

allenporter commented:
I've updated dialog text close to what I described above, then also added a `Learn about application credentials` link to the integration docs. Here is the updated view:

<img width="542" alt="Screen Shot 2022-09-04 at 2 03 38 PM" src="https://user-images.githubusercontent.com/6026418/188333469-68a9dc48-40e3-4c12-acc5-33597f05b1ca.png">

This has also been updated to reflect the changes in core, moving the logic for checking the config entry into a separate websocket API.

allenporter committed:
Fix lint errors and remove variables

matthiasdebaat commented:
As commented [here](https://github.com/home-assistant/frontend/issues/9773#issuecomment-1236922019). We should follow the guidelines of this is [great article](https://uxmag.com/articles/ui-copy-remove-vs-delete2-banner) about Delete vs. Remove. 

In this case we remove credentials from Home Assistant, they kept in existence on the service you've created the credentials on.

bramkragten commented on src/panels/config/integrations/ha-integration-card.ts:
```diff
@@ -716,6 +724,69 @@ export class HaIntegrationCard extends LitElement {
+            "ui.panel.config.integrations.config_entry.application_credentials.learn_more"
+          )}
+        </a>`,
+      confirmText: this.hass.localize("ui.common.remove"),
```
```suggestion
      destructive: true,
      confirmText: this.hass.localize("ui.common.remove"),
```

bramkragten commented on src/panels/config/integrations/ha-integration-card.ts:
```diff
@@ -716,6 +724,69 @@ export class HaIntegrationCard extends LitElement {
+      title: this.hass.localize(
+        "ui.panel.config.integrations.config_entry.application_credentials.delete_title"
+      ),
+      text: html` ${this.hass.localize(
```
```suggestion
      text: html`${this.hass.localize(
```

bramkragten commented on src/panels/config/integrations/ha-integration-card.ts:
```diff
@@ -716,6 +724,69 @@ export class HaIntegrationCard extends LitElement {
+  // if the integration isn't loaded or there is some other error.
+  private async _applicationCredentialForRemove(entryId: string) {
+    try {
+      return (await fetchApplicationCredentialsConfigEntry(this.hass, entryId))
```
will this always return an object?

allenporter replied:
I am not sure this is answering what you're looking for, but my impression is this will either:
- return a result with the `application_credentials_id` field when there are creds
- return an empty result in the common case where there are no creds, so `application_credentials_id` will not be in the response
- send an error of `invalid_config_entry_id`

My intent was that case 2 above would return undefined.

Let me know if there is a better pattern to follow.

bramkragten replied:
If case 2 returns undefined/null, we should make this:
```ts
      return (await fetchApplicationCredentialsConfigEntry(this.hass, entryId))?.application_credentials_id;
```

and update the typing

allenporter replied:
I was thinking that case 2 returns an empty message, so just the `application_credentials_id` field is undefined set:

```
export interface ApplicationCredentialsConfigEntry {
  application_credentials_id?: string;
}
```

Am i understanding that right?

bramkragten replied:
ok, in that case it is fine 👍

allenporter committed:
Apply suggestions from code review

Co-authored-by: Bram Kragten <mail@bramkragten.nl>

allenporter commented:
Backend PR was merged: https://github.com/home-assistant/core/pull/74825
